searx/engines/openstreetmap.py

```diff
@@ -23,24 +29,176 @@
 route_re = re.compile('(?:from )?(.+) to (.+)')
 
+wikidata_image_sparql = """
+select ?item ?p ?pLabel ?image ?itemLabel ?wikipediaLabel
+where {
+  values ?item { %QUERY% }
+  ?item wdt:P18|wdt:P1801|wdt:P2716|wdt:P3451|wdt:P4291|wdt:P4640|wdt:P4896|wdt:P5252|wdt:P5775|wdt:P8517 ?image .
+  SERVICE wikibase:label {
+    bd:serviceParam wikibase:language "[AUTO_LANGUAGE],en".
+    ?item rdfs:label ?itemLabel .
+  }
+  OPTIONAL { ?item wdt:P1766|wdt:P8505|wdt:P8667 ?sign }
+  OPTIONAL { ?item wdt:P41|wdt:P94|wdt:P154|wdt:P158|wdt:P2910|wdt:P4004|wdt:P5962|wdt:P8972 ?symbol }
+  OPTIONAL {
+    ?sitelink schema:about ?item;
+    schema:isPartOf <https://en.wikipedia.org/>;
+    schema:name ?wikipediaLabel .
+  }
+}
+ORDER by ?item ?p
+"""
+
+VALUE_TO_LINK = {
+    'phone': lambda v: 'tel:' + v,
+    'website': lambda v: v.split(';')[0],
+    'contact:website': lambda v: v.split(';')[0],
+    'email': lambda v: 'mailto:' + v,
+    'contact:email': lambda v: 'mailto:' + v,
+    'contact:phone': lambda v: 'tel:' + v,
+    'fax': lambda v: 'fax:' + v,
+    'contact:fax': lambda v: 'fax:' + v,
+    'contact:mastodon': lambda v: v,
+    'facebook': lambda v: v,
+    'contact:facebook': lambda v: v,
+    'contact:foursquare': lambda v: v,
+    'contact:instagram': lambda v: v,
+    'contact:linkedin': lambda v: v,
+    'contact:pinterest': lambda v: v,
+    'contact:telegram': lambda v: v,
+    'contact:tripadvisor': lambda v: v,
+    'contact:twitter': lambda v: v,
+    'contact:yelp': lambda v: v,
+    'contact:youtube': lambda v: v,
+    # 'mapillary': P1947
+    'contact:webcam': lambda v: v.split(';')[0],
+    'wikipedia': lambda v: 'https://{0}.wikipedia.org/wiki/{1}'.format(*v.split(':', 1)),
+    'wikidata': lambda v: 'https://wikidata.org/wiki/' + v,
+    'brand:wikidata': lambda v: 'https://wikidata.org/wiki/' + v,
+}
+KEY_ORDER = [
+    'cuisine',
+    'organic',
+    'delivery',
+    'delivery:covid19',
+    'opening_hours',
+    'opening_hours:covid19',
+    'fee',
+    'payment:*',
+    'currency:*',
+    'outdoor_seating',
+    'bench',
+    'wheelchair',
+    'level',
+    'building:levels',
+    'bin',
+    'public_transport',
+    'internet_access:ssid',
+]
+KEY_ORDER_INDEX = {k: i for i, k in enumerate(KEY_ORDER)}
+
+
+def get_key_rank(k):
+    ki = KEY_ORDER_INDEX.get(k)
+    if ki is None:
+        ki = KEY_ORDER_INDEX.get(k.split(':')[0] + ':*')
+    return ki
+
+
+def get_key(o):
+    return (get_key_rank(o['key']), o['label'])
+
 
 # do search-request
 def request(query, params):
 
-    params['url'] = base_url + search_string.format(query=query)
+    params['url'] = base_url + search_string.format(query=urlencode({'q': query}))
     params['route'] = route_re.match(query)
+    params['headers']['User-Agent'] = searx_useragent()
 
     return params
 
 
+def fetch_wikidata(nominatim_json, user_langage):
```

Comment:
I think functions like this needs to be documented .. for a newbee it will take hours and hours to understand whats going on here .. most often I gave up after 15min. 

The criticism is not related to this PR it is more general, I often stuck because no function in searx is documented. I don't think every little thing needs to be documented, but things that are not obvious from the implementation should be documented. 

What I mean: in the future we should more document .. I am also addressing myself :)

Reply:
Lines 131 to 138: 
* set `wikidata_id` to the wikidata ids of `['extratags']['wikidata']` or `['extratags']['wikidata link']` (I have seen both...)
* set `wd_to_results` to mapping from a wikidata id to a list of result (the actual result in `nominatim_json`).

* Line 142: send the query using `wikidata_id`.
* Line 144: for each result of the SPARQL request (more or less for each wikidata id (*) ).
* line 146: update the results of `nominatim_json` using `wd_to_results` = for each result of `nominatim_json` which reference `wikidata_id`, set the `wikidata` entry.